src/QsCompiler/CompilerOptimizations/CallableInlining.fs

```diff
@@ -0,0 +1,183 @@
+                // Only recurse if we haven't processed this callable yet
+                if found.Add ii.callable.FullName then
+                    findAllCalls callables ii.body found
+            | None -> ()
+        | _ -> ())
+
+/// Returns whether the given callable could eventually inline the given callable.
+/// Is used to prevent inlining recursive functions into themselves forever.
+let private cannotReachCallable (callables: Callables) (scope: QsScope) (cannotReach: QsQualifiedName) =
+    let mySet = HashSet()
+    findAllCalls callables scope mySet
+    not (mySet.Contains cannotReach)
+
+    
+/// The SyntaxTreeTransformation used to inline callables
+type internal CallableInliner(callables) =
+    inherit OptimizingTransformation()
+
+    // The current callable we're in the process of transforming
+    let mutable currentCallable: QsCallable option = None
+    // The list of new statements to add
+    let mutable statementsToAdd = []
+
+    /// Inline an expression as a ScopeStatement, with many checks to ensure correctness.
+    /// Returns None if the expression cannot be safely inlined.
+    let safeInline expr =
+        maybe {
+            let! ii = tryGetInliningInfo callables expr
+            
+            do! check (countReturnStatements ii.body = 0)
+            let! current = currentCallable
+            do! check (cannotReachCallable callables ii.body current.FullName)
+            do! check (ii.functors.controlled < 2)
+            
+            let newBinding = QsBinding.New ImmutableBinding (toSymbolTuple ii.specArgs, ii.arg)
+            let newStatements = ii.body.Statements.Insert (0, newBinding |> QsVariableDeclaration |> wrapStmt)
+            return {ii.body with Statements = newStatements} |> QsScopeStatement.New |> QsScopeStatement
+        }
+
+
+    override this.onCallableImplementation c =
+        let prev = currentCallable
+        currentCallable <- Some c
+        let result = base.onCallableImplementation c
+        currentCallable <- prev
+        result
+
+    override syntaxTree.Scope = { new ScopeTransformation() with
+
+        override scope.Expression = { new ExpressionTransformation() with
+            override expr.Kind = { new ExpressionKindTransformation() with 
+                override exprKind.ExpressionTransformation ex = expr.Transform ex
+                override exprKind.TypeTransformation t = expr.Type.Transform t
+
+                override exprKind.onOperationCall (method, arg) =
+                    let expr = CallLikeExpression (exprKind.ExpressionTransformation method, exprKind.ExpressionTransformation arg)
+                    maybe {
+                        let! scopeStatement = safeInline expr
+                        statementsToAdd <- statementsToAdd @ [wrapStmt scopeStatement]
+                        return UnitValue
+                    } |? expr
+
+                // The functions before are overriden to prevent inlining of lazily evaluated operations
```

Comment:
I am not sure I understand; the things you are skipping the transformation for are expressions - I was under then impression that this inlining routine would only consider inlining for expression statements?

Reply:
This is currently dead code that is there to support inlining callables inside expressions in the future (see earlier comment).

Reply:
Then I suggest to remove it before merging. Code should be useful or removed for the sake of maintainability.